src/IR/TypeInference.hs

```diff
@@ -4,72 +4,203 @@ For now, this module implements a very crude form of pure, syntax-directed type
+inferPrim (I.Prim I.Deref [e] _) = do
+  e' <- inferExpr e
+  return $ I.Prim I.Deref [e'] $ fromJust $ deref $ extract e'
+inferPrim (I.Prim I.Dup [e] _) = do
+  e' <- inferExpr e
+  return $ I.Prim I.Dup [e'] $ extract e'
+inferPrim e@(I.Prim I.Assign [lhs, rhs] _) = do
+  lhs' <- inferExpr lhs
+  rhs' <- inferExpr rhs
+  let rrty = ref $ extract rhs'
+  if extract lhs' == rrty
+    then do
+      case lhs' of
+        I.Var v _ -> insertVar v rrty
+        _ -> return ()
+      return $ I.Prim I.Assign [rrty <$ lhs', rhs'] unit
+    else throwError $ Compiler.TypeError $ "Assign expression has inconsistent type: " ++ show e
+inferPrim e@(I.Prim I.After [del, lhs, rhs] _) = do
+  del' <- inferExpr del
+  rhs' <- inferExpr rhs
+  lhs' <- inferExpr lhs
+  let rrty = ref $ extract rhs'
+  if extract del' == int 32 && extract lhs' == rrty
+    then do
+      case lhs' of
+        I.Var v _ -> insertVar v rrty
+        _ -> return ()
+      return $ I.Prim I.After [(\_ -> int 32) <$> del', rrty <$ lhs', rhs'] unit
+    else throwError $ Compiler.TypeError $ "After expression has inconsistent type: " ++ show e
+inferPrim (I.Prim (I.PrimOp I.PrimSub) [e1, e2]  _) = do
+  e1' <- inferExpr e1
+  e2' <- inferExpr e2
+  return $ I.Prim (I.PrimOp I.PrimSub) [e1', e2'] $ int 32
+inferPrim e@(I.Prim p [] _)
+  | p `elem` [I.Break, I.Return] = return $ I.Prim p [] void
+  | otherwise = throwError $ Compiler.TypeError $ "Unable to type Prim expression: " ++ show e
+inferPrim e@(I.Prim p es _)
+  | p `elem` [I.Loop, I.Wait, I.Par] = do
+      es' <- mapM inferExpr es
+      return $ I.Prim p es' unit
+  | otherwise = throwError $ Compiler.TypeError $ "Unable to type Prim expression: " ++ show e
+inferPrim e = throwError $ Compiler.TypeError $ "Unable to type Prim expression: " ++ show e
+
+localInferExpr :: [(Binder, I.Expr Ann.Type)] -> I.Expr Ann.Type -> InferFn ([(Binder, I.Expr Classes.Type)], I.Expr Classes.Type)
+localInferExpr [] body = do
+  body' <- inferExpr body
+  return ([], body')
+localInferExpr ((b, e):vs) body = do
+  e' <- inferExpr e
+  case b of
+    Just vid -> insertVar vid (extract e')
+    Nothing -> return ()
+  (vs', body') <- localInferExpr vs body
+  return ((b, e'):vs', body')
+
+withNewScope :: InferFn a -> InferFn a
+withNewScope inf = do
+  ctx <- get
+  x <- inf
+  modify $ const ctx
+  return x
+
+withVty :: I.Expr Ann.Type -> Binder -> Ann.Type -> InferFn ()
+withVty _ Nothing _ = return ()
+withVty e (Just v) t =
+  case dearrow t of
+    Just (vty, _) -> do
+      vty' <- anns2Class vty
+      insertVar v vty'
+    Nothing -> throwError $ Compiler.TypeError $ "Lambda wasn't annotated (or it wasn't an arrow type): " ++ show e
+
+anns2Class :: Ann.Type -> InferFn Classes.Type
+anns2Class (Ann.Type []) = throwError $ Compiler.TypeError "Cannot change empty Ann type to Classes type"
+anns2Class (Ann.Type ts) = ann2Class $ head ts
+
+ann2Class :: Ann.TypeAnnote -> InferFn Classes.Type
+ann2Class (Ann.TBuiltin bty) = case bty of
```

Comment:
Note to self: it seems like `Builtin t` is a `Functor` over `t` here, and we should be able to exploit this pattern.. maybe a `mapM` over the `Builtin` type.. (no action needed now but consider for future work.)

Reply:
Yeah we should discuss this more tomorrow!